File: countries/Countries.hs
module Countries where

import PGF

import Data.List
import Data.Char

data Country = Country {
  country :: String,
  capital :: String,
  area :: Int,
  population :: Int,
  continent :: String,
  currencyCode :: String,
  currencyName :: String
  }
  deriving Show


getCountries = do
  ts <- readFile "countries.tsv" >>= return . tail . fromTabs
  return [Country a b (readInt c) (readInt d) e f g | a:b:c:d:e:f:g:_ <- ts]

gfCountries = do
  cs <- getCountries
  writeFile "abs.tmp" $ unlines $ sort $ nub $ map (mkFunRule "Country" . country) cs
  appendFile "abs.tmp" $ unlines $ sort $ nub $ map (mkFunRule "Capital" . capital) cs
  appendFile "abs.tmp" $ unlines $ sort $ nub $ map (mkFunRule "Continent" . continent) cs
  appendFile "abs.tmp" $ unlines $ sort $ nub $ map (mkFunRule "CurrencyCode" . currencyCode) cs
  appendFile "abs.tmp" $ unlines $ sort $ nub $ map (mkFunRule "CurrencyName" . currencyName) cs
  writeFile "cnc.tmp" $ unlines $ sort $ nub $ map (mkLinRule "Country" . country) cs
  appendFile "cnc.tmp" $ unlines $ sort $ nub $ map (mkLinRule "Capital" . capital) cs
  appendFile "cnc.tmp" $ unlines $ sort $ nub $ map (mkLinRule "Continent" . continent) cs
  appendFile "cnc.tmp" $ unlines $ sort $ nub $ map (mkLinRule "CurrencyCode" . currencyCode) cs
  appendFile "cnc.tmp" $ unlines $ sort $ nub $ map (mkLinRule "CurrencyName" . currencyName) cs


readInt :: String -> Int
readInt s = if not (null s) && all isDigit s then (read s) else 0 ----

-----------------------

mkFunRule cat s = unwords ["fun", mkFunId cat s, ":", cat, ";"]
mkLinRule cat s = unwords ["lin", mkFunId cat s, "=", "mk" ++ cat, quoted s, ";"]

mkFunId cat s = showCId $ mkCId $ concat $ intersperse "_" $ words s ++ [cat]

quoted s = "\"" ++ s ++ "\""


-----------------------

fromTabs = map (getSep '\t') . lines
toTabs = unlines . map (blocks2sep "\t")

getSep :: Eq a => a -> [a] -> [[a]]
getSep sep xs = case break (==sep) xs of
  (x1,_:x2) -> x1 : getSep sep x2
  (x1,[]) -> x1 : []

blocks2sep :: String -> [String] -> String
blocks2sep sep = concat . intersperse sep

File: countries/countries.py
import pgf
from collections import namedtuple

def get_countries():
    countries = []
    Country = namedtuple('Country','country capital area population continent currencyCode currencyName')
    file = open('countries.tsv')
    lines = file.readlines()
    file.close()
    for line in lines[1:]:
        fields = Country(*line.split('\t'))
        countries.append(fields)
    return countries


def continent_article(countries,isin,cont):
  cocountries = [c for c in countries if isin(c)]
  return mkApp ("ContinentArticle", [
    constant("Continent",cont),
    mkInt(len(cocountries)),
    mkInt(sum([sint(c.area) for c in cocountries])),
    mkInt(sum([sint(c.population) for c in cocountries])),
    mkInt(average([sint(c.area) for c in cocountries])),
    mkInt(average([sint(c.population) for c in cocountries])),
    constant("Country", max(cocountries, key=lambda c: float(c.population)).country),
    constant("Country", min(cocountries, key=lambda c: float(c.population)).country),
    constant("Country", max(cocountries, key=lambda c: float(c.area)).country),
    constant("Country", min(cocountries, key=lambda c: float(c.area)).country)
    ])

     
def country_article(countr):
  return mkApp ("CountryArticle", [
    constant( "Country", countr.country),
    constant("Continent", countr.continent),
    constant("Capital", countr.capital),
    mkInt(countr.area),
    mkInt(countr.population),
    constant("CurrencyName", countr.currencyName),
    constant("CurrencyCode", countr.currencyCode)
    ])

def mkApp(f,xs):
    return pgf.Expr(f,xs)


def mkInt(s):
    return(pgf.readExpr(str(s)))


def constant(cat,f):
    return mkApp(mkFunId(cat,f),[])


def mkFunId(cat,fun):
    return '_'.join(fun.split() + [cat]) ###


def average(xs):
    return sum(xs)//len(xs)


def sint(s):
    if s.isdigit():
        return int(s)
    else:
        return round(float(s))


def main():
    countries = get_countries()
    continents = {c.continent for c in countries}
    
    worldtree = continent_article(countries, (lambda x: True), "world")
    continenttrees = [continent_article(countries, (lambda x: x.continent == c), c) for c in continents]
    countrytrees = [country_article(c) for c in countries]

    gr = pgf.readPGF("Countries.pgf")
    langs = list(gr.languages.values())
    langnum = 1

    for lang in langs:
        text = []
        text.append(lang.linearize(worldtree))
        for tree in continenttrees:
            text.append(lang.linearize(tree))
        for tree in countrytrees:
            text.append(lang.linearize(tree))
        file = open("countries_" + str(langnum) + ".html", "w")
        file.write('\n<p>'.join(text))
        file.close()
        langnum += 1



    
if __name__ == "__main__":
    main()






